torch/csrc/autograd/python_variable.cpp

```diff
@@ -914,30 +914,45 @@ int THPVariable_set_grad(THPVariable* self, PyObject* py_grad, void* unused) {
       "can't assign Variable as its own grad");
 
   const auto& grad = THPVariable_Unpack(py_grad);
-  TORCH_CHECK_TYPE(
+  TORCH_CHECK(
       var.dtype() == grad.dtype(),
       "attempting to assign a gradient with dtype '",
       grad.dtype(),
       "' to a tensor with dtype '",
       var.dtype(),
-      "'. Please ensure that the grad and the tensor have the same dtype");
-  bool gradIsSparse =
-      (var.dtype() == grad.dtype() &&
-       var.device().type() == grad.device().type() && grad.layout() == kSparse);
-  THPUtils_assertRet(
-      -1,
-      grad.options().type_equal(var.options()) || gradIsSparse,
-      "assigned grad has data of a different type");
+      "'. Please ensure that the gradient and the tensor have the same dtype");
+  TORCH_CHECK(
+      var.device().type() == grad.device().type(),
+      "attempting to assign a gradient with device type '",
+      grad.device().type(),
+      "' to a tensor with device type '",
+      var.device().type(),
+      "'. Please ensure that the gradient and the tensor are on the same device");
+  if (grad.layout() != kSparse) {
+    TORCH_CHECK(
+        var.layout() == grad.layout(),
+        "attempting to assign a gradient with ",
+        grad.layout(),
+        " layout to a tensor with ",
+        var.layout(),
+        " layout");
+  }
```

Comment:
If this is the way we want to go with this (see [here](https://github.com/pytorch/pytorch/pull/100683#issuecomment-1542683067) for context), I think it would probably be best to hardcode the message and remove the variable references. E.g., `grad.layout()` will always be 'Strided' and `var.layout()` will be 'Sparse'. I've kept it this way for now since I have an open question, as linked above

Reply:
will be going more general with this check, so this question is no longer relevant